### PR TITLE
Add a `jit` annotation around np.where.

### LINKED ISSUE #850: np.where performs broadcasting in op-by-op mode
The jax version of `np.where` needs to broadcast its arguments to the same shape. When one of the arguments is a constant, the broadcast operation isn't jitted.

Examples:
- `np.where(dropout_mask, x, 0)` creates an all-zero constant of the same shape as `x` and stores it in global memory. This potentially wastes hundreds of MB of global memory. 
- A similar problem occurs for the first argument (condition) of `np.where` if it has fewer dimensions than the other two arguments.

The first instance can be worked around by doing `np.where(condition, x, np.zeros_like(x))`, but working around the second is trickier without resorting to `lax.tie_in`.

### PR DESCRIPTION
**Pull request by hawkinsp**
Avoids materializing broadcast scalars inside where in op-by-op mode.

Since np.tril appears in the linear part of the Cholesky JVP rule, change np.tril/triu to avoid where in favor of calling lax.select() directly. Ban 1D arguments to np.tril/triu, which aren't a documented behavior of the numpy implementation.

Fixes #850

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
LGTM!

**Comment by nkitaev** [comment]
Does this approach fully address #850?

I've been looking into similar issues for dropout, and I think the following line:
https://github.com/google/jax/blob/623ddc373a9533b0fd25013db177a73b6aa3c1d3/jax/random.py#L425
gets executed in op-by-op mode when the dropout probability `p` is a python float, even though it's inside a jitted function. This leads me to believe that `@jit` around `np.where` might not solve the problem in all cases, especially when `np.where` itself is used inside a jit/pmap context.

**Comment by hawkinsp** [comment]
It certainly seems to. Note the absence of large constants in the following HLO IR:

```
In [1]: from jax import numpy as np; import jax

In [2]: print(jax.xla_computation(lambda x: np.where(x, 2, 3))(np.ones((1000,))).GetHloText())
HloModule jaxpr_computation__3.22

jaxpr_computation__4.9 {
  tuple.10 = () tuple()
  parameter.15 = s32[] parameter(4)
  parameter.16 = s32[] parameter(5)
  parameter.14 = f32[1000]{0} parameter(3)
  parameter.11 = f32[1000]{0} parameter(0)
  compare.17 = pred[1000]{0} compare(parameter.14, parameter.11), direction=NE
  parameter.12 = s32[] parameter(1)
  broadcast.18 = s32[1000]{0} broadcast(parameter.12), dimensions={}
  parameter.13 = s32[] parameter(2)
  broadcast.19 = s32[1000]{0} broadcast(parameter.13), dimensions={}
  ROOT select.20 = s32[1000]{0} select(compare.17, broadcast.18, broadcast.19)
}

ENTRY jaxpr_computation__3.22 {
  tuple.1 = () tuple()
  constant.2 = f32[] constant(0)
  broadcast.3 = f32[1000]{0} broadcast(constant.2), dimensions={}
  constant.4 = s32[] constant(2)
  constant.5 = s32[] constant(3)
  parameter.6 = f32[1000]{0} parameter(0)
  constant.7 = s32[] constant(2)
  constant.8 = s32[] constant(3)
  ROOT call.21 = s32[1000]{0} call(broadcast.3, constant.4, constant.5, parameter.6, constant.7, constant.8), to_apply=jaxpr_computation__4.9
}
```

If you have an example where it does not seem fixed for `np.where`, please share it!

We should probably look into the RNG issue separately, although we did have some ideas for a more expansive fix should they prove necessary.
